### DISCUSSION TITLE
Best way to managed who can sign in ? (with prisma)

### DISCUSSION THREAD

**Discussion by laurent512** [nextauthjs/next-auth#4855] [Help]
Hello,

First of all thank you for the great library, I am grateful for the amazing job you are doing. 

In my project, I enabled github providers with prisma adapter for authentication.
Issue is that I would love to be able to filter who can sign In based on their email address. 

I saw the callback method proposal and wondering what is the best  (safe & easy) way to filter who can sign in. 

```
async signIn({ user, account, profile, email, credentials }) {
      const isAllowedToSignIn = true;
      if (isAllowedToSignIn) {
        return true;
      } else {
        // Return false to display a default error message
        return false;
        // Or you can return a URL to redirect to:
        // return '/unauthorized'
      }
    },
```

For example, is it a working solution that in my User table of prisma, I add by myself record containing only the email of people allowed and verify in the signin callback if the email is present in that database ? 
Would that be a safe solution ? 
```
model User {
  id              String    @id @default(cuid())
  name            String?
  email           String?   @unique
  emailVerified   DateTime?
  image           String?
  accounts        Account[]
  sessions        Session[]
}
```

Thank you in advance !

**Comment by laurent512** [comment]
Any idea ? I would also appreciate any redirection ^^

**Reply by ThangHuuVu** [reply]
`signIn` is the intended place to achieve your goal. You could use the `userId` included in `user` or `account` to query the DB and check. Something like:
```ts
async signIn({ user, account, profile, email, credentials }) {
      const user = prisma.user.findUnique({
        where: {
          id: user.id
        }
      })
      if (user.email) {
        return true;
      } else {
        // Return false to display a default error message
        return false;
        // Or you can return a URL to redirect to:
        // return '/unauthorized'
      }
    },

```
